typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1005 @@
+
+```typescript
+type ArithExpr<A> =
+  | { type: "Num"; v: number; proof: Equality<number, A> }
+  | { type: "Plus"; l: ArithExpr<number>; r: ArithExpr<number>; proof: Equality<number, A> }
+  | { type: "Gt"; l: ArithExpr<number>; r: ArithExpr<number>; proof: Equality<boolean, A> }
+  | { type: "And"; l: ArithExpr<boolean>; r: ArithExpr<boolean>; proof: Equality<boolean, A> };
+```
+
+Now let's define our helper functions which help us to build expressions.
+And use `identity<A>(a: A): A` as a proof.
+
+```typescript
+const num: (v: number) => ArithExpr<number> = (v) => {
+  return { type: "Num", v, proof: identity };
+};
+
+const plus: (l: ArithExpr<number>, r: ArithExpr<number>) => ArithExpr<number> = (l, r) => {
+  return { type: "Plus", l, r, proof: identity };
+};
+
+const gt: (l: ArithExpr<number>, r: ArithExpr<number>) => ArithExpr<boolean> = (l, r) => {
+  return { type: "Gt", l, r, proof: identity };
+};
+
+const and: (l: ArithExpr<boolean>, r: ArithExpr<boolean>) => ArithExpr<boolean> = (l, r) => {
+  return { type: "And", l, r, proof: identity };
+};
+```
+
+The last step will be just `interpret` implementation.
+Switching on `expr.type` we construct our result running `interpret` on nested expressions and proofing resulted types with the proof from this `expr`.
+
+```typescript
+const interpret: <A>(expr: ArithExpr<A>) => A = (expr) => {
+  switch (expr.type) {
+    case "Num":
+      return expr.proof(expr.v);
+    case "Plus":
+      return expr.proof(interpret(expr.l) + interpret(expr.r));
+    case "Gt":
+      return expr.proof(interpret(expr.l) > interpret(expr.r));
+    case "And":
+      return expr.proof(interpret(expr.l) && interpret(expr.r));
+    default:
+      return expr;
+  }
+};
+
+const testExpr = and(gt(plus(num(23), num(12)), num(170)), gt(num(35), num(47)));
+interpret(testExpr); // false
+
+const wrongExpr = and(num(23), num(12)); // Argument of type 'ArithExpr<number>' is not assignable to parameter of type 'ArithExpr<boolean>'.
+```
+
+#### Tagless final eDSL
+
+Another way to implement eDSL is tagless final.
+Moving from data type to type class we are able to implement the same logic.
+
+Let's start from defining our type class.
+To make this we need already known `URIS` which is a union of all 1-arity defined in `URItoKind`.
+And also already known `Kind`.
+
+```typescript
+import { Kind, URIS } from "fp-ts/HKT";
+
+type ArithExpr<Expr extends URIS> = {
+  num: (v: number) => Kind<Expr, number>;
+  plus: (l: Kind<Expr, number>, r: Kind<Expr, number>) => Kind<Expr, number>;
+  gt: (l: Kind<Expr, number>, r: Kind<Expr, number>) => Kind<Expr, boolean>;
+  and: (l: Kind<Expr, boolean>, r: Kind<Expr, boolean>) => Kind<Expr, boolean>;
+};
+```
+
+Now, we will present our types for which we will create `ArithExpr` instances.
+The first one will be just interpreter.
+And the second one will create string representation of expression.
```

Comment:
You can insert the types here, I think. 

We need the description to be as maximally close to the code as possible.